### ISSUE TITLE
some parameters appear in more than one parameter group

### ISSUE THREAD

**Issue by quancs** [speediedan/finetuning-scheduler#4]
## 🐛 Bug

The following code reports the error: some parameters appear in more than one parameter group.
Comments case 1, and uncomments case 2, the error disappears.

### To Reproduce

```python
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, LightningDataModule
from pytorch_lightning.utilities.cli import LightningArgumentParser, LightningCLI


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer1 = torch.nn.Linear(32, 32)
        self.layer2 = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer2(self.layer1(x))

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.Adam(self.parameters(), lr=0.1)
        # case 1: with scheduler
        self.scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizer, factor=0.5, patience=5)
        return {
            'optimizer': optimizer,
            'lr_scheduler': {
                'scheduler': self.scheduler,
                'monitor': 'valid_loss',
            }
        }
        # case 2: without scheduler
        # return optimizer


class MyDataModule(LightningDataModule):

    def __init__(self):
        super().__init__()

    def train_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def val_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def test_dataloader(self) -> DataLoader:
        return DataLoader(RandomDataset(32, 64), batch_size=2)


from finetuning_scheduler import FinetuningScheduler, FTSCheckpoint, FTSEarlyStopping


class BoringCLI(LightningCLI):

    def add_arguments_to_parser(self, parser: LightningArgumentParser) -> None:
        parser.add_lightning_class_args(FinetuningScheduler, 'finetune_scheduler')
        parser.set_defaults({
            'finetune_scheduler.ft_schedule': {
                0: {
                    'params': ['layer1'],  # the parameters for each phase definition can be fully specified
                    'max_transition_epoch': 2,
                    'lr': 0.001,
                },
                1: {
                    'params': ['layer2'],
                    'lr': 0.001 
                }
            },
        })

        # EarlyStopping
        parser.add_lightning_class_args(FTSEarlyStopping, "early_stopping")
        early_stopping_defaults = {
            "early_stopping.monitor": "valid_loss",
            "early_stopping.patience": 99999,  # disable early_stopping
            "early_stopping.mode": "min",
            "early_stopping.min_delta": 0.01,
        }
        parser.set_defaults(early_stopping_defaults)

        # ModelCheckpoint
        parser.add_lightning_class_args(FTSCheckpoint, 'model_checkpoint')
        model_checkpoint_defaults = {
            "model_checkpoint.filename": "epoch{epoch}_val_loss{valid_loss:.2f}",
            "model_checkpoint.monitor": "valid_loss",
            "model_checkpoint.mode": "min",
            "model_checkpoint.every_n_epochs": 1,
            "model_checkpoint.save_top_k": 5,
            "model_checkpoint.auto_insert_metric_name": False,
            "model_checkpoint.save_last": True
        }
        parser.set_defaults(model_checkpoint_defaults)


if __name__ == "__main__":
    BoringCLI(BoringModel, MyDataModule, seed_everything_default=None, save_config_overwrite=True)

```

### Environment

- Fine-Tuning Scheduler Version (e.g., 0.1.0): 0.2.0
- PyTorch Lightning Version (e.g., 1.5.0):  1.7.0
- PyTorch Version (e.g., 1.10): 1.12.1
- Python version (e.g., 3.9): 3.9.7
- OS (e.g., Linux): ubuntu 22.04
- CUDA/cuDNN version: 11.6
- GPU models and configuration: NVIDIA A100-SXM4-80GB
- How you installed PyTorch (`conda`, `pip`, source): conda
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

**Comment by speediedan** [comment]
Glad to see you trying out Fine-Tuning Scheduler! 

The root of the issue you describe is that all parameters are being passed in the creation of the optimizer, including those that were set to be frozen by FTS in accordance with the schedule (i.e. 'layer2' params were configured to be frozen but still passed to the initial optimizer). While instantiation of the optimizer is left the user, your issue has led me realize FTS could be improved in by issuing a warning to the user in cases like this where parameters with ``requires_grad=False`` are included in the initial optimizer instantiation but also referenced in subsequent training phases. In the meantime, the issue above could be addressed with a ``requires_grad`` filter applied to the optimizer instantiation like so:
 
```python
    def configure_optimizers(self):
        # optimizer = torch.optim.Adam(self.parameters(), lr=0.1)
        parameters = filter(lambda x: x.requires_grad, self.parameters())
        optimizer = torch.optim.Adam(parameters, lr=0.1)
        ...
```
BTW, I also noticed in the example above you're using a scheduler that's not a subclass of ``torch.optim._LRScheduler``. Though the current release of FTS [indicates this scheduler requirement](https://finetuning-scheduler.readthedocs.io/en/stable/advanced/lr_scheduler_reinitialization.html#:~:text=Currently%2C%20all%20_LRScheduler%20s%20are%20supported%20with%20the%20exception%20of%20ChainedScheduler%20and%20SequentialLR%20) in the lr scheduler reinitialization documentation section, the check for this requirement is only done for schedules using lr reinitialization. For the next FTS patch release, I'll include a fix to mention the scheduler requirement in the primary FTS documentation (not just the lr reinitialization section) and to check for supported scheduler use in the ``on_fit_start`` callback hook. In the meantime, you can use any of the other [lr schedulers](https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate) that are [supported by FTS](https://finetuning-scheduler.readthedocs.io/en/stable/advanced/lr_scheduler_reinitialization.html#:~:text=Currently%2C%20all%20_LRScheduler%20s%20are%20supported%20with%20the%20exception%20of%20ChainedScheduler%20and%20SequentialLR%20), so for instance, this will work: 

```python
        ...
        # case 1: with scheduler
        # self.scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizer, factor=0.5, patience=5)
        self.scheduler = torch.optim.lr_scheduler.StepLR(optimizer=optimizer, step_size=1, gamma=0.7)
        ...
```
Thanks so much for the valuable feedback! Your usage elucidated two ways in which FTS could be improved and I think commits addressing both will be included in the next patch release. 

I hope you find FTS somewhat helpful in your research and I'm always happy to help answer any questions... love the increasingly collaborative and open-source nature of the ML research community! Feel free to reach out anytime, I love hearing if and how researchers are using FTS in their experimentation. Wish you all the best with your future work!

**Comment by quancs** [comment]
Thank you for your replies.

According to you advice, I changed the `configure_optimizers` to:
```python
    def configure_optimizers(self):
        parameters = filter(lambda x: x.requires_grad, self.parameters())
        optimizer = torch.optim.Adam(parameters, lr=0.1)
        # case 1: with scheduler
        self.scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer=optimizer, factor=0.5, patience=5)
        self.scheduler.base_lrs = []
        return {
            'optimizer': optimizer,
            'lr_scheduler': {
                'scheduler': self.scheduler,
                'monitor': 'valid_loss',
            }
        }
```
where, `self.scheduler.base_lrs = []` is to used to suppress the error `'ReduceLROnPlateau' object has no attribute 'base_lrs'
`. I know it must have some side effects like the lrs specified in the config will not work.

The reason I want to use `ReduceLROnPlateau` is that the paper I'm trying to compare with (on a new dataset which is not the one used in that paper) uses this lr_scheduler. Changing `ReduceLROnPlateau` to other lr_schedulers will cause the reproducing imperfect. So, is there any way to support `ReduceLROnPlateau` and any other user-defined lr_schedulers? For example, give the user a way to know the current phase number, and then let the user to adjust their lr_scheduler by themself. Or, let the user provide a function which accepts the lr specified in config and their lr_scheduler instance, and executing the function will set the lr_scheduler properly. 

It would be much better to have a way to support that. I willing to help if there is anything I can do.

**Comment by speediedan** [comment]
Totally understand the reproduciblity requirement and agree 100% on the need for ``ReduceLROnPlateau`` support. 

``ReduceLROnPlateau`` wasn't supported in the first minor release of FTS because of the non-trivial incremental complexity in supporting its divergent configuration in the context of both implicit and explicit mode fine-tuning (it's different from the ``_LRScheduler``s in some important ways). While one could subclass ``_LRScheduler`` (like is done for a [custom LambdaLR example here](https://github.com/speediedan/finetuning-scheduler/blob/a66345ffbd9dd6228d5ced149fc057615285ae4c/tests/helpers/boring_model.py#L21-L30)) and replicate the functionality of ``ReduceLROnPlateau`` that's not the right solution and I wouldn't ask you to do that :rofl: 

Instead,  I've experimented today with implementing support for ``ReduceLROnPlateau`` directly and I think it'll be doable without adding *too* much complexity. I've got a few things going in parallel but I'll try to bang out beta support ASAP and let you know as soon as it's pushed (it should land in the next patch release not too long after that).

As far as granting the user the flexibility to set ``lr``s arbitrarily in different phases, that should be achievable in both implicit and explicit fine-tuning scheduler modes using one or more [LambdaLR schedulers](https://finetuning-scheduler.readthedocs.io/en/stable/advanced/lr_scheduler_reinitialization.html#:~:text=Phases%20with%20defined%20lr%20scheduler%20reinitialization%20configs%20will%20always%20apply%20the%20specified%20config%2C%20including%20new%20lambdas%20if%20provided). You can always both specify the desired scheduler for a given phase and the initial ``lr``  (or initial ``lr`` for [each individual parameter group](https://finetuning-scheduler.readthedocs.io/en/stable/advanced/lr_scheduler_reinitialization.html#:~:text=%3A%20Explicit_Reinit_LR_Scheduler%0A18-,init_pg_lrs%3A%20%5B2.0e%2D06%2C%202.0e%2D06%5D,-Phase%202%20in) in explicit fine-tuning mode).

Thanks again for your valuable feedback, I'm definitely open and eager to hear suggestions on improvements and PRs are always welcome!  Thanks again for your thoughts, your feedback is already improving FTS in a couple ways @quancs!

**Comment by quancs** [comment]
Great! I will try the next release as soon as it comes.

**Comment by speediedan** [comment]
A new [FTS patch release (0.2.1)](https://finetuning-scheduler.readthedocs.io/en/v0.2.1/) is out now including several of the potential improvements you helped elucidate, thanks @quancs! I think the new release (specifically 0a3b2b88155d1af569652f3ddb0c64078ae30a93, 3b33c7f3671675bb38376ca0a43e8bcd3ec24d1e, aa6399ed2f28c3d8902a43e17545c19781f3ec26) should close this issue now but feel free to open other issues as necessary or reach out anytime. Thanks again for your valuable feedback and good luck with your future research!

**Comment by quancs** [comment]
Nice, I will try it.
